components/settings/settings.cpp

```diff
@@ -117,6 +120,230 @@ class SettingsFileParser
+
+            // Settings not in the written map are definitely invalid.  Currently, this can only
+            // happen if the player edited the file while playing, because loadSettingsFile()
+            // will accept anything and pass it along in the map, but in the future, we might
+            // want to handle invalid settings more gracefully here.
+            if (finder == written.end()) {
+                ostream << "# invalid setting: " << line << std::endl;
+                changed = true;
+                continue;
+            }
+
+            // Write the current value of the setting to the file.  The key must exist in the
+            // settings map because of how written was initialized and finder != end().
+            ostream << setting << " = " << settings[key] << std::endl;
+            // Mark that setting as written.
+            finder->second = true;
+            // Did we really change it?
+            if (value != settings[key]) {
+                std::cout << "Changed setting: [" << currentCategory << "] "
+                          << setting << " = " << settings[key] << std::endl;
+                changed = true;
+            }
+            // No need to write the current line, because we just emitted a replacement.
+
+            // Curiously, it appears that comments at the ends of lines with settings are not
+            // allowed, and the comment becomes part of the value.  Was that intended?
+        }
+
+        // We're done with the input stream file.
+        istream.close();
+
+        // Ensure that all options in the current category have been written.  We must complete
+        // the current category at the end of the file before moving on to any new categories.
+        for (CategorySettingStatusMap::iterator mit = written.begin(); mit != written.end(); ++mit) {
+            bool missed = false;
+            if (mit->second == false && mit->first.first == currentCategory) {
+                std::cout << "Added new setting: [" << mit->first.first << "] "
+                          << mit->first.second << " = " << settings[mit->first] << std::endl;
+                // With some further code changes, this comment could be more descriptive.
+                ostream << std::endl << "# Automatically added setting." << std::endl;
+                ostream << mit->first.second << " = " << settings[mit->first] << std::endl;
+                mit->second = true;
+                missed = true;
+                changed = true;
+            }
+            // Ensure that there's a newline before the next section if we added settings.
+            if (missed) ostream << std::endl;
+        }
+
+        // This logic triggers when the input file was effectively empty.  It could be extended
+        // to doing something more intelligent instead, like make a copy of the default settings
+        // file (complete with comments) before continuing.  Other code prevents OpenMW from
```

Comment:
> ran a build of openmw-launcher without my changes. After loading a save file and immediately exiting, my user settings.cfg contained 121 settings in 16 sections. :-o

Was not aware of that. Definitely not intended. One more reason to nuke the launcher settings code and make it use the existing settings component.

Try removing your settings file, then running openmw directly, not through the launcher. That should yield the intended behaviour. 

> Also, following your recent removal of "ui sensitivity", I received assertion style errors refusing to run OpenMW because my settings.cfg did not contain the required value.

If the settings.cfg does not contain a required value, we fall back to settings-default.cfg. In this case, I have to assume you pulled the latest code, but didn't actually compile it - so the setting was removed, but the code accessing the setting was still there. In such cases throwing a runtime error is the correct thing to do.

> That's the approach I was designing for, but not because I preferred it -- because that's how it seems to work now.
> 
> Should I understand your comments here to mean that the goal is to reverse that behavior?

Yes.

As far as I could tell, your branch <i>does</i> behave correct w.r.t. not copying the complete default settings into user settings (barring the launcher bug). I just mentioned the issue because the comment on this line suggested that as a future possibility:

> It could be extended
+        // to doing something more intelligent instead, like make a copy of the default settings
+        // file (complete with comments) before continuing